### PR TITLE
*: Support modify table/column charset from latin1 to utf8/utf8mb4

### LINKED ISSUE #34008: Unsupported modify charset from latin1 to utf8mb4
## Bug Report

<p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">This can be easily simulated<br>MySQL<span> </span><span class="error">[test]</span>&gt; create table t2(a varchar(20) )CHARSET=latin1;<br>MySQL<span> </span><span class="error">[test]</span>&gt; insert into t2 values('&lt;some value&gt;');<br>MySQL<span> </span><span class="error">[test]</span>&gt; alter table t2 DEFAULT CHARACTER SET utf8mb4;<br>ERROR 8200 (HY000): Unsupported modify charset from latin1 to utf8mb4</p><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">If we create another table with charset=utf8mb4 and insert from t2. There is no error<br>MySQL<span> </span><span class="error">[test]</span>&gt; create table t3 (a varchar(20)) charset=utf8mb4;<br>MySQL<span> </span><span class="error">[test]</span>&gt; insert into t3 select * from t2;<br>Query OK, 1 row affected (0.01 sec)<br>MySQL<span> </span><span class="error">[test]</span>&gt; select * from t3;<br><ins>------</ins></p><div class="table-wrap" style="margin: 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

a
--

</div><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><ins>------</ins><br>It is not clear if the above insert would convert latin char to utf8 correctly.</p><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Customer is aware of the latin support issue as documented and in<span> </span><a href="https://github.com/pingcap/tidb/issues/18955" class="external-link" target="_blank" rel="nofollow noopener" style="color: rgb(0, 82, 204); text-decoration: var(--aui-link-decoration); cursor: pointer;">https://github.com/pingcap/tidb/issues/18955</a></p>

### 1. Minimal reproduce step (Required)
MySQL [test]> create table t2(a varchar(20) )CHARSET=latin1;
MySQL [test]> insert into t2 values('<some value>');
MySQL [test]> alter table t2 DEFAULT CHARACTER SET utf8mb4;

### 2. What did you expect to see? (Required)
Query OK
### 3. What did you see instead (Required)
ERROR 8200 (HY000): Unsupported modify charset from latin1 to utf8mb4
### 4. What is your TiDB version? (Required)
v5.2.2

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: ref #34008 
Problem Summary: new feature for modify charset from latin1 to utf8/utf8mb4.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
* Let latin1 as an alias for utf8 in tidb
* Support modify table/column charset from latin1 to utf8/utf8mb4
```

### CODE REVIEW

**Review by tangenta — APPROVED** [review]

File: ddl/db_integration_test.go [hunk 1]
```diff
@@ -769,9 +769,15 @@ func TestChangingTableCharset(t *testing.T) {
 	tk.MustGetErrCode("alter table t charset utf8 collate utf8mb4_bin;", errno.ErrCollationCharsetMismatch)
 	tk.MustGetErrCode("alter table t charset utf8 collate utf8_bin collate utf8mb4_bin collate utf8_bin;", errno.ErrCollationCharsetMismatch)
 
-	tk.MustGetErrCode("alter table t charset utf8", errno.ErrUnsupportedDDLOperation)
-	tk.MustGetErrCode("alter table t charset utf8mb4", errno.ErrUnsupportedDDLOperation)
-	tk.MustGetErrCode("alter table t charset utf8mb4 collate utf8mb4_bin", errno.ErrUnsupportedDDLOperation)
+	tk.MustExec("alter table t charset utf8")
```
Comment (zimulala) [inline ddl/db_integration_test.go:772]: Please add the check of `admin check table t`.
And do we need to add some records?

File: cmd/explaintest/t/collation_misc.test [hunk 2]
```diff
@@ -12,12 +12,25 @@ create table t(a varchar(20) charset latin1);
 insert into t values ("t_value");
 alter table t modify column a varchar(20) charset latin1;
 select * from t;
---error 8200
+
 alter table t modify column a varchar(20) charset utf8;
```
Comment (zimulala) [inline cmd/explaintest/t/collation_misc.test:16]: Please add the `admin check table t`.
Add please do this for other similar places.

File: cmd/explaintest/t/collation_misc.test [hunk 3]
```diff
@@ -12,12 +12,25 @@ create table t(a varchar(20) charset latin1);
...
---error 8200
+select * from t;
+
+drop table t;
+create table t(a varchar(20) charset latin1);
+insert into t values ("t_value");
 alter table t modify column a varchar(20) charset utf8mb4;
```
Comment (zimulala) [inline cmd/explaintest/t/collation_misc.test:22]: Do we need to add tests that change column types with reorg in this PR?

File: ddl/db_integration_test.go (hunk 1 above)
Comment (Defined2014) [reply ddl/db_integration_test.go:772]: Done. I think with data is covered by `collation_misc.test`

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**